File: python/3sum.py
from typing import List


class Solution:
    def threeSum(self, nums: List[int]) -> List[List[int]]:
        nums = sorted(nums)
        answer = set()
        l = len(nums)
        for i in range(l - 2):
            if i > 0 and nums[i] == nums[i - 1]:
                continue
            left = i + 1
            right = l - 1

            while left < right:
                s = nums[i] + nums[left] + nums[right]
                if s > 0:
                    right -= 1
                elif s < 0:
                    left += 1
                else:
                    answer.add((nums[i], nums[left], nums[right]))
                    left += 1
                    right -= 1

        answer = [list(a) for a in answer]
        return answer


s = Solution()
print(s.threeSum([-1, 0, 1, 2, -1, -4]))
print(s.threeSum([]))
print(s.threeSum([0]))
print(s.threeSum([0, 0, 0]))

File: c/3_sum.cpp
#include <stdio.h>
#include <stdlib.h>

void swap(int* nums, int left, int right) {
	int temp = nums[left];
	nums[left] = nums[right];
	nums[right] = temp;
}

void qsort(int* nums, int pivot, int left, int right) {
	int end = right;
	if (left > right) {
		return;
	}

	while (left < right) {
		if (nums[left] > nums[right]) {
			swap(nums, left, right);
		}
			left++;
			right--;	
	}
	swap(nums, pivot, right);
	qsort(nums, pivot, pivot + 1, right - 1);
	qsort(nums, left, left, end);
}

int** threeSum(int* nums, int numsSize, int* returnSize, int** returnColumnSizes) {
	qsort(nums, 0, 1, numsSize - 1);
	int maxCandidate = numsSize * (numsSize - 1) / 2;
	int** returnArray = (int**)malloc(sizeof(int*) * 3000);
	return returnArray;
}

int main(void) {
	int nums1[3000];

	int nums[5] = { 5, 4, 3, 2, 1 };
	threeSum(nums, 5, 0, 0);
}
